### PR TITLE
Feat/compute sn address

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #397 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/kakarot_core.cairo [hunk 1]
```diff
@@ -158,10 +165,32 @@ mod KakarotCore {
...
         fn compute_starknet_address(
             self: @ContractState, evm_address: EthAddress
         ) -> ContractAddress {
-            panic_with_felt252('not implemented')
+            // Deployer is always Kakarot Core
+            let deployer = get_contract_address();
+
+            // Constructor Calldata
+            // For an EOA, the constructor calldata is:
+            // fn constructor(kakarot_address: ContractAddress, evm_address: EthAddress)
```
Comment (enitrat) [inline crates/evm/src/kakarot_core.cairo:177]:
```suggestion
            // [kakarot_address, evm_address]
```

File: crates/evm/src/kakarot_core.cairo [hunk 2]
```diff
@@ -158,10 +165,32 @@ mod KakarotCore {
...
+
+            let hash = PedersenTrait::new(0)
+                .update(CONTRACT_ADDRESS_PREFIX)
+                .update_with(deployer)
+                .update_with(evm_address)
+                .update_with(self.eoa_class_hash.read())
+                .update(constructor_calldata_hash)
+                .finalize();
+
+            let normalized_address: Option<ContractAddress> = (hash.into() & POW_2_251).try_into();
```
Comment (enitrat) [inline crates/evm/src/kakarot_core.cairo:191]:
```suggestion
            let normalized_address:ContractAddress = (hash.into() & POW_2_251).try_into().unwrap();
```

Directly

File: crates/utils/src/traits.cairo [hunk 3]
```diff
@@ -44,6 +44,16 @@ impl EthAddressIntoU256 of Into<EthAddress, u256> {
     }
 }
 
+impl U256TryIntoContractAddress of TryInto<u256, ContractAddress> {
+    fn try_into(self: u256) -> Option<ContractAddress> {
+        let intermediate: Option<felt252> = self.try_into();
+        match intermediate {
+            Option::Some(felt) => felt.try_into(),
```
Comment (enitrat) [inline crates/utils/src/traits.cairo:51]:
```suggestion
            Option::Some(value) => value.try_into(),
```

avoid using data types names for variables

File: crates/utils/src/traits.cairo [hunk 4]
```diff
@@ -44,6 +44,16 @@ impl EthAddressIntoU256 of Into<EthAddress, u256> {
     }
 }
 
+impl U256TryIntoContractAddress of TryInto<u256, ContractAddress> {
+    fn try_into(self: u256) -> Option<ContractAddress> {
+        let intermediate: Option<felt252> = self.try_into();
```
Comment (enitrat) [inline crates/utils/src/traits.cairo:49]:
```suggestion
        let maybe_value: Option<felt252> = self.try_into();
```

File: crates/utils/src/traits.cairo [hunk 5]
```diff
@@ -44,6 +44,16 @@ impl EthAddressIntoU256 of Into<EthAddress, u256> {
...
+        match intermediate {
```
Comment (enitrat) [inline crates/utils/src/traits.cairo:50]:
```suggestion
        match maybe_value {
```

**Review by enitrat — COMMENTED** [review]
lgtm

**Comment by Eikix** [comment]
Thanks for the review siirrrr 🥛🥛🍼🧋

I need a bit more time to write tests for it!

**Comment by enitrat** [comment]
bigger tests 🏋️

**Review by enitrat — APPROVED** [review]

**Review by enitrat — COMMENTED** [review]
remove multiplication

File: crates/core_contracts/src/kakarot_core.cairo [hunk 6]
```diff
@@ -122,7 +122,9 @@ mod KakarotCore {
 
     #[derive(Drop, starknet::Event)]
     struct EOADeployed {
+        #[key]
```
Comment (enitrat) [inline crates/core_contracts/src/kakarot_core.cairo:125]: why two keys and no data?

File: crates/core_contracts/src/kakarot_core.cairo [hunk 7]
```diff
@@ -183,16 +185,18 @@ mod KakarotCore {
             // For an EOA, the constructor calldata is:
             // [kakarot_address, evm_address]
             let constructor_calldata_hash = PedersenTrait::new(0)
-                .update(deployer.into())
-                .update(evm_address.into())
+                .update_with(deployer)
+                .update_with(evm_address)
+                .update(2)
```
Comment (enitrat) [inline crates/core_contracts/src/kakarot_core.cairo:190]: what was the reason? can you document it somewhere?

File: crates/core_contracts/src/kakarot_core.cairo [hunk 8]
```diff
@@ -183,16 +185,18 @@ mod KakarotCore {
...
                 .finalize();
 
             let hash = PedersenTrait::new(0)
-                .update(CONTRACT_ADDRESS_PREFIX)
+                .update_with(CONTRACT_ADDRESS_PREFIX)
```
Comment (enitrat) [inline crates/core_contracts/src/kakarot_core.cairo:194]: if it's a felt - why not `update`?

I don't specifically mind just want to know if theres' a reason behind

File: crates/core_contracts/src/kakarot_core.cairo [hunk 9]
```diff
@@ -246,7 +250,7 @@ mod KakarotCore {
             self.eoa_address_registry.write(evm_address, starknet_address);
 
             // Emit an event
-            self.emit(EOADeployed { evm_address, starknet_address });
+            self.emit(EOADeployed { evm_address, starknet_address: starknet_address });
```
Comment (enitrat) [inline crates/core_contracts/src/kakarot_core.cairo:253]: why hard naming?
If required, simply do `:starknet_address` as both the field and the variable share the same name

File: crates/core_contracts/src/tests/test_kakarot_core.cairo [hunk 10]
```diff
@@ -77,31 +78,29 @@ fn test_kakarot_core_set_native_token() {
 fn test_kakarot_core_deploy_eoa() {
     let kakarot_core = utils::deploy_kakarot_core();
     let eoa_starknet_address = kakarot_core.deploy_eoa(test_utils::evm_address());
+    // We drop the first even of Kakarot Core, as it is the initializer from Ownable,
```
Comment (enitrat) [inline crates/core_contracts/src/tests/test_kakarot_core.cairo:81]:
```suggestion
    // We drop the first event of Kakarot Core, as it is the initializer from Ownable,
```

File: crates/core_contracts/src/kakarot_core.cairo (hunk 8 above)
Comment (Eikix) [reply crates/core_contracts/src/kakarot_core.cairo:194]: no special reason, just to minimally use felts linked API as much as possible

File: crates/core_contracts/src/kakarot_core.cairo (hunk 6 above)
Comment (Eikix) [reply crates/core_contracts/src/kakarot_core.cairo:125]: i've heard keys are easier to index, that way in an indexor, one could search per starknet address and search per evm address?

Comment (enitrat) [reply crates/core_contracts/src/kakarot_core.cairo:125]: ok

**Review by enitrat — APPROVED** [review]
lgtm

**Pull request merged**